### PR TITLE
[6.0][ClangImporter] Make sure the `-resource-dir` is checked before the `-sdk`, as done everywhere else in the compiler

### PR DESCRIPTION
**Pull request by finagolfin**
__Explanation:__ The frontend currently looks in `-sdk` for a C/C++ sysroot and `-resource-dir` for Swift modules, with the former implicitly set to `/` and the latter always explicitly passed in by `swift-driver`. However, a "full SDK" contains both C/C++ headers _and_ Swift modules, so in that case and _if_ `-sdk` is explicitly passed in, this `ClangImporter` was looking in the full SDK first and wrongly using those Swift module maps, rather than the ones in `-resource-dir`.

__Scope:__ When does that actually happen? Suppose you have Swift 5.10 installed in the system and are trying to build Swift 6.0 from source. In that scenario, the CMake config explicitly passes `-sdk /` and [the Swift 6 compiler wrongly uses the old Swift 5.10 C and C++ module maps and fails](https://github.com/swiftlang/swift/issues/74696#issuecomment-2198714412). This could also affect some SDK bundles, say if one linux distro triple's Swift modules are placed in the common SDK itself, while others are put in separate `-resource-dir` paths, since compiling SDK bundles passes in an explicit `-sdk`.

It _does not_ affect simply building Swift packages with a local SwiftPM 6.0 while the implicit system SDK path of `/` has a system Swift 5.10 installed, as no explicit `-sdk` is passed to build those packages.

__Issue:__ #74696

__Original PRs:__ #74814 and #76119

__Risk:__ Low, since this only affects the few listed scenarios where the `-sdk` flag is explicitly passed to the compiler

__Testing:__ Passed all CI on trunk and my and others' testing locally. @compnerd did report an issue with his new Android SDK with an earlier version of this patch, so I'll wait on his further testing and approval before asking to merge this.

__Reviewer:__ @egorzhdan

### CODE REVIEW

**Review by egorzhdan — COMMENTED** [review]
This might be a breaking change for certain setups, I don't think we can land this kind of change in a bugfix release.

**Comment by finagolfin** [comment]
> This might be a breaking change for certain setups

Such as?

> I don't think we can land this kind of change in a bugfix release.

Ideally, we'd get it into Swift 6.0.0 also, once it's accepted into this branch.

**Comment by DougGregor** [comment]
> > This might be a breaking change for certain setups
> 
> Such as?

I think the answer is that we don't know, but past experience has shown that changing search paths can break configurations that we hadn't heard about. 

This does look like the right set of changes to make, but I'm somewhat concerned that the issue @compnerd hit with https://github.com/swiftlang/swift/pull/74814 isn't known to be resolved.

**Comment by finagolfin** [comment]
> I think the answer is that we don't know, but past experience has shown that changing search paths can break configurations that we hadn't heard about.

OK, I looked into it and Saleem added this first SDK lookup five years ago in 370aa9d9c, precisely to look for these module maps in full SDKs that are causing the problems now. I think this was to paper over the frontend not looking for Swift modules in full SDKs altogether, as I reported earlier this year in swiftlang/swift-driver#1562, and @artemcm's Frontend pull for that is how it should've been fixed instead.

I don't think most compilation scenarios set the `-sdk` flag, as I noted in the writeup about `Risk`, so the only question is whether this is worth fixing for the niche scenarios I listed where the current lookup order is broken for the potential unknown niche scenarios that depend on the current lookup order.

> I'm somewhat concerned that the issue @compnerd hit with https://github.com/swiftlang/swift/pull/74814 isn't known to be resolved.

That was with an earlier version of this patch, but since this merged version was made available a couple weeks ago, their CI has been broken or they have been traveling, so we have not received any notice on this patch.

I suspect this patch works fine, but as I said above, I'm still waiting on them to verify.

**Comment by finagolfin** [comment]
@shahmishal, we're going on three weeks with no response here, can we get this in? I suspect the reason Saleem had an issue with an earlier version of this patch is that I made a mistake initially, which Egor then found and we corrected.

This has been in trunk for a month now without any reported issues and [distros like Fedora have to pull it in for their 6.0 build](https://src.fedoraproject.org/rpms/swift-lang/blob/ea96108046f8a1021f3bbd46d95c8cef7799626e/f/resource_dir.patch).

**Comment by DougGregor** [comment]
@swift-ci please test

**Comment by finagolfin** [comment]
Ping @shahmishal or @DougGregor, this passed the 6.0 CI, just need a decision now.

**Comment by finagolfin** [comment]
Ping, this has been in trunk for four months now and in the downstream 6.0 release of Fedora for the last month: I think we have a pretty good body of evidence that it is unlikely to break anything.

**Comment by finagolfin** [comment]
Ping @DougGregor, with the next patch release coming up, can you make a decision on this? I consider this a small fix for more advanced build scenarios, but I leave it up to you release managers if it's worth backporting.

**Review by DougGregor — APPROVED** [review]
Yeah, we should take this. Thank you!

**Pull request merged**
